evalml/pipelines/components/transformers/preprocessing/stl_decomposer.py

```diff
@@ -0,0 +1,347 @@
+
+        To emulate the sklearn API, when the STL decomposer is fit, the full seasonal
+        component, a single period sample of the seasonal component, the full
+        trend-cycle component and the residual are saved.
+
+        y(t) = S(t) + T(t) + R(t)
+
+        Args:
+            X (pd.DataFrame, optional): Conditionally used to build datetime index.
+            y (pd.Series): Target variable to detrend and deseasonalize.
+
+        Returns:
+            self
+
+        Raises:
+            ValueError: If y is None.
+            ValueError: If target data doesn't have DatetimeIndex AND no Datetime features in features data
+        """
+        X, y = self._check_target(X, y)
+
+        # Warn for poor decomposition use with higher periods
+        if self.seasonal_period > 14:
+            str_dict = {"D": "daily", "M": "monthly"}
+            data_str = ""
+            if y.index.freqstr in str_dict:
+                data_str = str_dict[y.index.freqstr]
+            self.logger.warning(
+                f"STLDecomposer may perform poorly on {data_str} data with a high seasonal period ({self.seasonal_period}).",
+            )
+
+        # Save the frequency of the fitted series for checking against transform data.
+        self.frequency = y.index.freqstr
+
+        stl = STL(y, seasonal=self.seasonal_period)
+        res = stl.fit()
+        self.seasonal = res.seasonal
+        self.seasonality = self.seasonal[: self.seasonal_period]
+        self.trend = res.trend
+        self.residual = res.resid
+
+        return self
+
+    def transform(
+        self,
+        X: pd.DataFrame,
+        y: pd.Series = None,
+    ) -> tuple[pd.DataFrame, pd.Series]:
+        """Transforms the target data by removing the polynomial trend and rolling average seasonality.
+
+        Uses an ARIMA model to project forward the addititve trend and removes it. Then, utilizes the first period's
+        worth of seasonal data determined in the .fit() function to extrapolate the seasonal signal of the data to be
+        transformed.  This seasonal signal is also assumed to be additive and is removed.
+
+        Args:
+            X (pd.DataFrame, optional): Conditionally used to build datetime index.
+            y (pd.Series): Target variable to detrend and deseasonalize.
+
+        Returns:
+            tuple of pd.DataFrame, pd.Series: The input features are returned without modification. The target
+                variable y is detrended and deseasonalized.
+
+        Raises:
+            ValueError: If target data doesn't have DatetimeIndex AND no Datetime features in features data
+        """
+        if y is None:
+            return X, y
+        X, y = self._check_target(X, y)
+
+        self._check_oos_past(y)
+
+        y_in_sample = pd.Series([])
+        y_out_of_sample = pd.Series([])
+
+        # For wholly in-sample data, retrieve stored results.
+        if y.index[0] <= self.trend.index[-1] and y.index[0] >= self.trend.index[0]:
```

Comment:
should this be `if y.index[-1] <= self.trend.index[-1]` so that y is in the range of the trend? might just be my misunderstanding with the stuff above as well..